lib/client.js

```diff
@@ -1607,41 +1605,61 @@
     return this.__performSystemRequest('revoke-token', null, null, callback);
   };
 
-  HappnClient.prototype.__endSocket = function() {
-    this.socket.removeAllListeners('end');
-    this.socket.end();
-    delete this.socket;
+  HappnClient.prototype.__destroySocket = function(socket, callback) {
+    if (socket._local)
+      //this is an eventemitter connection
+      return setTimeout(() => {
+        socket.end();
+        callback();
+      }, 0);
+    if (socket.readyState !== 1) {
+      socket.end().destroy();
+      return callback();
+    }
+    socket
+      .once('close', () => {
+        // wait for next loop cycle to allow for primus cleanup
+        setTimeout(() => {
+          let destroyError;
+          try {
+            socket.destroy();
+          } catch (e) {
+            this.log.warn('socket.destroy failed in client: ' + e.toString());
+            destroyError = e;
+          }
+          callback(destroyError);
+        }, 0);
+      })
+      .end();
   };
 
-  HappnClient.prototype.__connectionCleanup = function(options) {
-    try {
-      if (!this.socket) return;
-      if (!options) options = {};
-      var _this = this;
+  HappnClient.prototype.__connectionCleanup = function(options, callback) {
+    if (typeof options === 'function') {
+      callback = options;
+      options = null;
+    }
 
-      if (options.revokeToken || options.revokeSession)
-        return this.revokeToken(function(e) {
-          if (e)
-            _this.log.warn('socket.end failed in client, revoke-token failed: ' + e.toString());
-          _this.__endSocket();
-        });
+    if (!this.socket) return callback();
+    if (!options) options = {};
 
-      if (options.disconnectChildSessions)
-        return this.__performSystemRequest('disconnect-child-sessions', null, null, function(e) {
-          if (e)
-            _this.log.warn(
-              'socket.end failed in client, disconnect-child-sessions failed: ' + e.toString()
-            );
-          _this.__endSocket();
-        });
+    if (options.revokeToken || options.revokeSession)
+      return this.revokeToken(e => {
+        if (e) this.log.warn('socket.end failed in client, revoke-token failed: ' + e.toString());
+        this.__destroySocket(this.socket, callback);
+      });
 
-      this.__endSocket();
-    } catch (e) {
-      this.log.warn('socket.end failed in client: ' + e.toString());
-    }
+    if (options.disconnectChildSessions)
+      return this.__performSystemRequest('disconnect-child-sessions', null, null, e => {
+        if (e)
+          this.log.warn(
+            'socket.end failed in client, disconnect-child-sessions failed: ' + e.toString()
```

Comment:
is socket.end the correct context message here?

Reply:
fixed